go/worker/compute/executor/committee/node.go

```diff
@@ -581,74 +576,81 @@ func (n *Node) HandleNewBlockLocked(blk *block.Block) {
 			"round", blk.Header.Round,
 		)
 
-		n.scheduleCh.In() <- struct{}{}
+		n.checkTxCh.In() <- struct{}{}
 	}
 }
 
-// checkTx requests the runtime to check the validity of the given transaction.
-func (n *Node) checkTx(ctx context.Context, rawTx []byte) (*transaction.CheckedTransaction, error) {
+// checkTxs requests the runtime to check the validity of the given transaction batch.
+func (n *Node) checkTxs() {
+	batch := n.checkTxQueue.GetBatch()
+	if len(batch) == 0 {
+		return
+	}
+	rt := n.GetHostedRuntime()
+	if rt == nil {
+		n.logger.Error("CheckTx: hosted runtime not initialized")
+		return
+	}
+
 	n.commonNode.CrossNode.Lock()
 	currentBlock := n.commonNode.CurrentBlock
 	currentConsensusBlock := n.commonNode.CurrentConsensusBlock
 	currentEpoch := n.commonNode.Group.GetEpochSnapshot().GetEpochNumber()
 	n.commonNode.CrossNode.Unlock()
 
-	rt := n.GetHostedRuntime()
-	if rt == nil {
-		n.logger.Error("CheckTx: hosted runtime not initialized")
-		return nil, errNotReady
+	// Check transaction batch.
+	results, err := rt.CheckTx(n.ctx, currentBlock, currentConsensusBlock, currentEpoch, batch)
+	if err != nil {
+		n.logger.Error("transaction batch check tx error", "err", err)
+		return
 	}
 
-	tx, err := rt.CheckTx(ctx, currentBlock, currentConsensusBlock, currentEpoch, rawTx)
-	switch {
-	case err == nil:
-	case errors.Is(err, host.ErrInvalidArgument):
-		return nil, errNotReady
-	case errors.Is(err, host.ErrInternal):
-		return nil, err
-	case errors.Is(err, host.ErrCheckTxFailed):
-		return nil, p2pError.Permanent(err)
-	default:
-		return nil, err
+	txs := make([]*transaction.CheckedTransaction, 0, len(results))
+	for i, res := range results {
+		if !res.IsSuccess() {
+			n.logger.Warn("check tx failed", "tx", batch[i], "result", res)
+			continue
+		}
+
+		txs = append(txs, res.ToCheckedTransaction(batch[i]))
 	}
-	return tx, nil
+
+	// Remove the checked transaction batch.
+	n.checkTxQueue.RemoveBatch(batch)
+
+	// Queue checked transactions for scheduling.
+	n.queueTxs(txs)
 }
 
-// QueueTx queues a runtime transaction for scheduling.
-func (n *Node) QueueTx(tx *transaction.CheckedTransaction) error {
+// queues a runtime transaction batch for scheduling.
+func (n *Node) queueTxs(txs []*transaction.CheckedTransaction) {
 	n.schedulerMutex.RLock()
 	defer n.schedulerMutex.RUnlock()
 
-	if n.scheduler == nil {
-		return errNotReady
-	}
-
-	// Check if transaction was recently scheduled.
-	if n.lastScheduledCache != nil {
-		if _, b := n.lastScheduledCache.Get(tx.Hash()); b {
-			return errDuplicateTx
+	for _, tx := range txs {
+		// Skip recently scheduled transactions.
```

Comment:
Maybe this check should be performed before even checking the batch (e.g. in `checkTxBatch`)?

Reply:
Yeah could do that, probably even before, in `HandlePeerMessage`. Not sure if we should also move the insert there? or only insert it here, once the transaction has passed the check?

Currently the transactions that fail CheckTx are missed by this cache - however a transaction could fail a CheckTx but pass it a future time. So probably better to keep the insertion here, only for transactions that pass the check.

Reply:
Yes, I think it makes sense to move the cache check into `HandlePeerMessage`, but insertion should be kept here after the transaction passes CheckTx for the reasons you mentioned above.